feat: add StructuredData handler to support inserting "application/ld+json" scripts

smerchek opened the pull request:
This PR adds a new `<StructuredData />` handler, specifically for "application/ld+json" scripts (https://developers.google.com/search/docs/advanced/structured-data/intro-structured-data).

There are a couple new opinions in this PR to improve typings, that we can discuss (I'll comment in the relevant code sections).

I've also add `schema-dts` as a dev dependency, in order to give StructuredData really good types. Let me know if you'd prefer to leave these types to user to define.

sergiodxa reviewed (commented):
A question, is it ok to have more than one script tag with structured data? Shouldn't a single page have a single script with ld+json type?

From what I saw here https://developers.google.com/search/docs/advanced/structured-data/sd-policies#nesting it's a single script tag with an array of values.

If that's the case, the `key` is also not required, the component should get all the structured data objects and inline a the array instead of rendering multiple script tags.

smerchek commented:
> A question, is it ok to have more than one script tag with structured data? Shouldn't a single page have a single script with ld+json type?
> 
> From what I saw here https://developers.google.com/search/docs/advanced/structured-data/sd-policies#nesting it's a single script tag with an array of values.
> 
> If that's the case, the `key` is also not required, the component should get all the structured data objects and inline a the array instead of rendering multiple script tags.

You are right! Historically, our code had included multiple scripts tags per item. However, while that may be supported, it seems that it's not recommended and makes this simpler. So, I'll get that updated.

smerchek commented:
@sergiodxa I think this is in a good place now. Let me know if there is anything else that should be addressed.

sergiodxa commented on README.md:
```diff
@@ -260,6 +260,78 @@ Now, any script you defined in the ScriptsFunction will be added to the HTML tog
+import type { WithContext, BlogPosting } from 'schema-dts';
+
+// export the handle with the correct type:
+export let handle: HandleStructuredData<LoaderData> = {
```
What do you think if you pass a second generic to `HandleStructuredData` where you can send `BlogPosting`? And use that to ensure the return type of `StructuredDataFunction` is a `WithContext<BlogPosting>` type. That way you can return directly inside the function without assigning to a variable.

A thing to consider is what happens when there's no data, the return type of `StructuredDataFunction` could be `WithContext<Result> | null` and return `null` when you can't return anything.

If you want to return an array instead of an object I suppose you could do `HandleStructuredData<LoaderData, BlogPosting[]>` and that should work?

smerchek replied:
`null` is a simple case to support.

I like the idea of making the `Thing` type explicit, if you want. The single element is easy to type, but I'll have to work with supporting the array well. My use-case of `BreadcrumbList` + `BlogPosting` would need to specify two types with the resulting type being `[WithContext<BreadcrumbList>, WithContext<BlogPosting>]`.

sergiodxa commented on README.md:
```diff
@@ -260,6 +260,78 @@ Now, any script you defined in the ScriptsFunction will be added to the HTML tog
+        {children}
+        <ScrollRestoration />
+        <Scripts />
+        <StructuredData />
```
Is it ok to render this at the end of the `<body>`? Shouldn't it be inside `<head>`?

sergiodxa commented on README.md:
````diff
@@ -260,6 +260,78 @@ Now, any script you defined in the ScriptsFunction will be added to the HTML tog
+}
+```
+
+Now, any structured data you defined in the StructuredDataFunction will be added to the HTML, in the body.
````
```suggestion
Now, any structured data you defined in the `StructuredDataFunction` will be added to the HTML, in the body.
```

sergiodxa commented on src/react/structured-data.tsx:
```diff
@@ -0,0 +1,91 @@
+ * @example
+ * // This route uses the data to render structured data (e.g. BreadcrumbList and BlogPosting)
+ * export let handle = {
+ *    structuredData: (data: LoaderData) => {
```
another nit:
```suggestion
 *    structuredData(data: LoaderData) {
```

sergiodxa commented:
@smerchek a few more questions and ideas, but it looks good, thanks for working on this!

smerchek commented:
Ok @sergiodxa 

Please take another look.

I added the optional second type param, support `null` being returned, and I also added some tests for a few of the scenarios. 

As part of the tests, I fixed the existing react.test.tsx and added tsx file matchers to the jest config. Maybe there is a better way to use the jsdom environment for the react tsx test files?
